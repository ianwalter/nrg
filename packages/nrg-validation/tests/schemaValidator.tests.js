import { test } from '@ianwalter/bff'
import {
  isString,
  isEmail,
  isPhone,
  isStrongPassword,
  isOptional,
  lowercase,
  trim,
  SchemaValidator
} from '../index.js'

const containsSoftware = {
  validate: occupation => ({
    isValid: occupation.toLowerCase().includes('software'),
    message: 'Occupation must contain software.'
  })
}
const registrationValidator = new SchemaValidator({
  email: { isEmail, lowercase, trim },
  name: { isString },
  password: { isStrongPassword, message: 'Your password must be stronger.' },
  occupation: { containsSoftware },
  phone: { isPhone, isOptional, name: 'telephone number' }
})
const email = 'yo@fastmail.com'
const validInput = {
  email: '  yo@FASTmail.com',
  name: 'Georgy Zhukov',
  password: '23-01=dwko;qwe2',
  occupation: 'Software General',
  phone: '555-555-5555'
}

test('valid registration', async ({ expect }) => {
  const validation = await registrationValidator.validate(validInput)
  expect(validation.isValid).toBe(true)
})

test('invalid registration', async ({ expect }) => {
  const input = {
    email: 'hahaha',
    name: '',
    password: 'qwerty',
    occupation: 'CEO',
    phone: '777'
  }
  const validation = await registrationValidator.validate(input)
  expect(validation.isValid).toBe(false)
  expect(validation).toMatchSnapshot({
    validations: {
      password: {
        result: expect.any(Object)
      }
    }
  })
})

test('validaion data', async ({ expect }) => {
  const input = {
    ...validInput,
    artist: 'Peach Pit',
    song: 'Feelin Low'
  }
  const validation = await registrationValidator.validate(input)
  expect(validation.data).toEqual({ ...validInput, email })
})

test('without optional data', async ({ expect }) => {
  const { phone, ...required } = validInput
  const validation = await registrationValidator.validate(required)
  expect(validation.isValid).toBe(true)
})
