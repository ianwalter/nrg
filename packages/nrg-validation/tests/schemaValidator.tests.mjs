import { test } from '@ianwalter/bff'
import {
  isString,
  isBoolean,
  isInteger,
  isArray,
  isEmail,
  isPhone,
  isStrongPassword,
  isOptional,
  lowercase,
  trim,
  SchemaValidator,
  ignoreEmpty
} from '../index.js'

const containsSoftware = {
  validate: occupation => ({
    isValid: occupation.toLowerCase().includes('software'),
    message: 'Occupation must contain software.'
  })
}
const registrationValidator = new SchemaValidator({
  email: { isEmail, lowercase, trim },
  name: isString,
  password: { isStrongPassword, message: 'Your password must be stronger.' },
  occupation: { containsSoftware },
  phone: { isPhone, isOptional, name: 'telephone number' },
  organizationName: { isString, ignoreEmpty }
})
const email = 'yo@fastmail.com'
const validInput = {
  email: '  yo@FASTmail.com',
  name: 'Georgy Zhukov',
  password: '23-01=dwko;qwe2',
  occupation: 'Software General',
  phone: '6177779501',
  organizationName: 'Acme'
}
const args = { phone: ['US'] }

test('Valid registration', async t => {
  const validation = await registrationValidator.validate(validInput, args)
  t.expect(validation.isValid).toBe(true)
})

test('Invalid registration', async t => {
  const input = {
    email: 'hahaha',
    name: '',
    password: 'qwerty',
    occupation: 'CEO',
    phone: '777',
    organizationName: false
  }
  const validation = await registrationValidator.validate(input, args)
  t.expect(validation.isValid).toBe(false)
  t.expect(validation).toMatchSnapshot({
    validations: {
      password: {
        result: t.expect.any(Object)
      }
    }
  })
})

test('With validation data', async t => {
  const input = {
    ...validInput,
    artist: 'Peach Pit',
    song: 'Feelin Low'
  }
  const validation = await registrationValidator.validate(input, args)
  t.expect(validation.data).toEqual({ ...validInput, email })
})

test('Without optional data', async t => {
  const { phone, ...required } = validInput
  const validation = await registrationValidator.validate(required, args)
  t.expect(validation.isValid).toBe(true)
})

test('Field with ignoreEmpty', async t => {
  const input = { ...validInput, organizationName: '' }
  const validation = await registrationValidator.validate(input, args)
  t.expect(validation.isValid).toBe(true)
})

test('Nested SchemaValidator', async t => {
  const orderValidator = new SchemaValidator({
    userId: { isInteger },
    consent: { isBoolean },
    cart: new SchemaValidator({
      currency: { isString },
      products: { isArray: isArray(isString) }
    })
  })
  const validTop = { userId: 123, consent: true }
  const validCart = { currency: 'USD', products: ['widget', 'gizmo'] }

  // Test validating valid input.
  let validation = await orderValidator.validate({
    ...validTop,
    cart: validCart,
    coupon: 'FREE'
  })
  t.expect(validation.isValid).toBe(true)
  t.expect(validation.data.coupon).toBe(undefined)

  // Test validating invalid top-level input.
  validation = await orderValidator.validate({ userId: 123, cart: validCart })
  t.expect(validation.isValid).toBe(false)
  t.expect(validation.validations.userId.isValid).toBe(true)
  t.expect(validation.validations.consent.isValid).toBe(false)

  // Test validating invalid nested input.
  const invalidCart = { currency: 'GBP', products: ['widget', { qty: 1 }] }
  validation = await orderValidator.validate({ ...validTop, cart: invalidCart })
  t.expect(validation.isValid).toBe(false)
})
