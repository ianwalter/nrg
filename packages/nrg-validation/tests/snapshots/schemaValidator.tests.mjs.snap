// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid registration 1`] = `
Object {
  "data": Object {
    "email": "hahaha",
    "name": "",
    "occupation": "CEO",
    "password": "qwerty",
    "phone": "777",
  },
  "feedback": Object {
    "email": Array [
      "A valid email is required.",
    ],
    "name": Array [
      "A valid name is required.",
    ],
    "occupation": Array [
      "Occupation must contain software.",
    ],
    "password": Array [
      "This is a top-10 common password",
      "Add another word or two. Uncommon words are better.",
    ],
  },
  "isValid": false,
  "validations": Object {
    "email": Object {
      "isValid": false,
    },
    "name": Object {
      "isValid": false,
    },
    "occupation": Object {
      "isValid": false,
      "message": "Occupation must contain software.",
    },
    "password": Object {
      "feedback": Array [
        "Add another word or two. Uncommon words are better.",
      ],
      "isValid": false,
      "message": "This is a top-10 common password",
      "result": Any<Object>,
    },
  },
}
`;
